File: server/quest-bot.ts
import { Client, GatewayIntentBits, REST, Routes, SlashCommandBuilder, EmbedBuilder, PermissionsBitField } from 'discord.js';
import mysql from 'mysql2/promise';

// Database connection configuration - using same config as main app
const dbConfig = {
  host: process.env.DB_HOST || 'db2.sillydevelopment.co.uk',
  user: process.env.DB_USER || 'u77272_CezJ7ZJDoG',
  password: process.env.DB_PASSWORD || '4R.u8LGwD10VjCh84af=k4Vh',
  database: process.env.DB_NAME || 's77272_axiom',
  port: parseInt(process.env.DB_PORT || '3306'),
};

// Create Discord client for second bot
const client = new Client({
  intents: [
    GatewayIntentBits.Guilds,
    GatewayIntentBits.GuildMessages,
    GatewayIntentBits.MessageContent,
    GatewayIntentBits.GuildMembers,
  ],
});

// Database connection pool
let db: mysql.Pool;

// Bot settings storage
const guildSettings = new Map<string, {
  questClaimChannelId?: string;
  boostChannelId?: string;
  prefix?: string;
}>();

// Quest notification commands
const commands = [
  new SlashCommandBuilder()
    .setName('setquestchannel')
    .setDescription('Set the channel for quest completion notifications')
    .addChannelOption(option =>
      option.setName('channel')
        .setDescription('Channel to send quest notifications')
        .setRequired(true)
    ),

  new SlashCommandBuilder()
    .setName('setboostchannel')
    .setDescription('Set the channel for boost notifications')
    .addChannelOption(option =>
      option.setName('channel')
        .setDescription('Channel to send boost notifications')
        .setRequired(true)
    ),

  new SlashCommandBuilder()
    .setName('removequestchannel')
    .setDescription('Remove quest notification channel'),

  new SlashCommandBuilder()
    .setName('removeboostchannel')
    .setDescription('Remove boost notification channel'),

  new SlashCommandBuilder()
    .setName('questsettings')
    .setDescription('View current quest notification settings'),

  new SlashCommandBuilder()
    .setName('setprefix')
    .setDescription('Set command prefix for page information commands')
    .addStringOption(option =>
      option.setName('prefix')
        .setDescription('New command prefix (e.g., "!")')
        .setRequired(true)
    ),
];

// Initialize database connection
async function initDatabase() {
  try {
    db = mysql.createPool(dbConfig);
    console.log('✅ Connected to external MySQL database');
    
    // Test connection
    const connection = await db.getConnection();
    await connection.ping();
    connection.release();
    console.log('✅ Database connection verified');
  } catch (error) {
    console.error('❌ Failed to connect to database:', error);
  }
}

// Register slash commands
async function registerCommands() {
  try {
    const rest = new REST({ version: '10' }).setToken(process.env.BOT2_TOKEN!);
    
    console.log('Started refreshing quest bot application (/) commands.');

    await rest.put(
      Routes.applicationCommands(client.user!.id),
      { body: commands.map(command => command.toJSON()) },
    );

    console.log('Successfully reloaded quest bot application (/) commands.');
  } catch (error) {
    console.error('Error registering quest bot commands:', error);
  }
}

client.once('clientReady', async () => {
  console.log(`✅ Quest Bot logged in as ${client.user?.tag}!`);
  await initDatabase();
  await registerCommands();
});

// Handle slash commands
client.on('interactionCreate', async (interaction) => {
  if (!interaction.isChatInputCommand()) return;

  const { commandName } = interaction;

  try {
    switch (commandName) {
      case 'setquestchannel':
        await handleSetQuestChannel(interaction);
        break;
      case 'setboostchannel':
        await handleSetBoostChannel(interaction);
        break;
      case 'removequestchannel':
        await handleRemoveQuestChannel(interaction);
        break;
      case 'removeboostchannel':
        await handleRemoveBoostChannel(interaction);
        break;
      case 'questsettings':
        await handleQuestSettings(interaction);
        break;
      case 'setprefix':
        await handleSetPrefix(interaction);
        break;
    }
  } catch (error) {
    console.error('Error handling quest bot command:', error);
    if (!interaction.replied) {
      await interaction.reply({ content: 'An error occurred while processing your command.', ephemeral: true });
    }
  }
});

// Command handlers
async function handleSetQuestChannel(interaction: any) {
  if (!interaction.member.permissions.has(PermissionsBitField.Flags.ManageChannels)) {
    await interaction.reply({ content: '❌ You need **Manage Channels** permission to set quest channels.', ephemeral: true });
    return;
  }

  const channel = interaction.options.getChannel('channel');
  const guildId = interaction.guild.id;

  const settings = guildSettings.get(guildId) || {};
  settings.questClaimChannelId = channel.id;
  guildSettings.set(guildId, settings);

  await interaction.reply({
    embeds: [{
      title: '✅ Quest Channel Set',
      description: `Quest completion notifications will now be sent to ${channel}`,
      color: 0x00ff00,
      timestamp: new Date().toISOString()
    }]
  });
}

async function handleSetBoostChannel(interaction: any) {
  if (!interaction.member.permissions.has(PermissionsBitField.Flags.ManageChannels)) {
    await interaction.reply({ content: '❌ You need **Manage Channels** permission to set boost channels.', ephemeral: true });
    return;
  }

  const channel = interaction.options.getChannel('channel');
  const guildId = interaction.guild.id;

  const settings = guildSettings.get(guildId) || {};
  settings.boostChannelId = channel.id;
  guildSettings.set(guildId, settings);

  await interaction.reply({
    embeds: [{
      title: '✅ Boost Channel Set',
      description: `Server boost notifications will now be sent to ${channel}`,
      color: 0x00ff00,
      timestamp: new Date().toISOString()
    }]
  });
}

async function handleRemoveQuestChannel(interaction: any) {
  if (!interaction.member.permissions.has(PermissionsBitField.Flags.ManageChannels)) {
    await interaction.reply({ content: '❌ You need **Manage Channels** permission to remove quest channels.', ephemeral: true });
    return;
  }

  const guildId = interaction.guild.id;
  const settings = guildSettings.get(guildId) || {};
  
  if (!settings.questClaimChannelId) {
    await interaction.reply({ content: '❌ No quest channel is currently set.', ephemeral: true });
    return;
  }

  delete settings.questClaimChannelId;
  guildSettings.set(guildId, settings);

  await interaction.reply({
    embeds: [{
      title: '✅ Quest Channel Removed',
      description: 'Quest completion notifications have been disabled.',
      color: 0xff6b6b,
      timestamp: new Date().toISOString()
    }]
  });
}

async function handleRemoveBoostChannel(interaction: any) {
  if (!interaction.member.permissions.has(PermissionsBitField.Flags.ManageChannels)) {
    await interaction.reply({ content: '❌ You need **Manage Channels** permission to remove boost channels.', ephemeral: true });
    return;
  }

  const guildId = interaction.guild.id;
  const settings = guildSettings.get(guildId) || {};
  
  if (!settings.boostChannelId) {
    await interaction.reply({ content: '❌ No boost channel is currently set.', ephemeral: true });
    return;
  }

  delete settings.boostChannelId;
  guildSettings.set(guildId, settings);

  await interaction.reply({
    embeds: [{
      title: '✅ Boost Channel Removed',
      description: 'Server boost notifications have been disabled.',
      color: 0xff6b6b,
      timestamp: new Date().toISOString()
    }]
  });
}

async function handleQuestSettings(interaction: any) {
  const guildId = interaction.guild.id;
  const settings = guildSettings.get(guildId) || {};

  const embed = new EmbedBuilder()
    .setTitle('🎯 Quest Bot Settings')
    .setColor(0x7c3aed)
    .addFields(
      {
        name: '🎉 Quest Channel',
        value: settings.questClaimChannelId ? `<#${settings.questClaimChannelId}>` : 'Not set',
        inline: true
      },
      {
        name: '🚀 Boost Channel',
        value: settings.boostChannelId ? `<#${settings.boostChannelId}>` : 'Not set',
        inline: true
      },
      {
        name: '⚙️ Command Prefix',
        value: settings.prefix || '!' ,
        inline: true
      }
    )
    .setTimestamp();

  await interaction.reply({ embeds: [embed] });
}

async function handleSetPrefix(interaction: any) {
  if (!interaction.member.permissions.has(PermissionsBitField.Flags.ManageGuild)) {
    await interaction.reply({ content: '❌ You need **Manage Server** permission to set the command prefix.', ephemeral: true });
    return;
  }

  const prefix = interaction.options.getString('prefix');
  const guildId = interaction.guild.id;

  if (prefix.length > 3) {
    await interaction.reply({ content: '❌ Prefix must be 3 characters or less.', ephemeral: true });
    return;
  }

  const settings = guildSettings.get(guildId) || {};
  settings.prefix = prefix;
  guildSettings.set(guildId, settings);

  await interaction.reply({
    embeds: [{
      title: '✅ Prefix Updated',
      description: `Command prefix is now set to: \`${prefix}\`\n\nTry: \`${prefix}page home\` or \`${prefix}page servers\``,
      color: 0x00ff00,
      timestamp: new Date().toISOString()
    }]
  });
}

// Handle prefix commands for page information
client.on('messageCreate', async (message) => {
  if (message.author.bot) return;

  const guildId = message.guild?.id;
  if (!guildId) return;

  const settings = guildSettings.get(guildId);
  const prefix = settings?.prefix || '!';

  if (!message.content.startsWith(prefix)) return;

  const args = message.content.slice(prefix.length).trim().split(/ +/);
  const command = args.shift()?.toLowerCase();

  if (command === 'page') {
    await handlePageCommand(message, args);
  }
});

// Page information command handler
async function handlePageCommand(message: any, args: string[]) {
  if (args.length === 0) {
    await message.reply('Usage: `!page <page_name>` (e.g., `!page home`, `!page servers`, `!page bots`)');
    return;
  }

  const pageName = args[0].toLowerCase();
  const baseUrl = process.env.APP_URL || 'https://smartserve.repl.co';

  try {
    // Fetch page information from website
    const response = await fetch(`${baseUrl}/api/page-info/${pageName}`);
    
    if (!response.ok) {
      await message.reply(`❌ Page "${pageName}" not found. Available pages: home, servers, bots, store, quests`);
      return;
    }

    const pageInfo = await response.json();

    const embed = new EmbedBuilder()
      .setTitle(`📄 ${pageInfo.title || 'Page Information'}`)
      .setDescription(pageInfo.description || 'No description available')
      .setColor(0x7c3aed)
      .setURL(`${baseUrl}${pageInfo.url || '/'}`)
      .setTimestamp();

    if (pageInfo.features && pageInfo.features.length > 0) {
      embed.addFields({
        name: '✨ Features',
        value: pageInfo.features.join('\n'),
        inline: false
      });
    }

    if (pageInfo.stats) {
      embed.addFields({
        name: '📊 Statistics',
        value: Object.entries(pageInfo.stats)
          .map(([key, value]) => `**${key}**: ${value}`)
          .join('\n'),
        inline: false
      });
    }

    await message.reply({ embeds: [embed] });
  } catch (error) {
    console.error('Error fetching page info:', error);
    await message.reply('❌ Failed to fetch page information. Please try again later.');
  }
}

// Quest completion notification function
export async function sendQuestNotification(guildId: string, userId: string, questData: {
  questId: string;
  questName: string;
  reward: number;
  userTag: string;
  newBalance: number;
}) {
  try {
    const settings = guildSettings.get(guildId);
    if (!settings?.questClaimChannelId) return;

    const channel = await client.channels.fetch(settings.questClaimChannelId);
    if (!channel || !('send' in channel)) return;

    const embed = new EmbedBuilder()
      .setTitle('🎉 Quest Completed!')
      .setDescription(`**${questData.userTag}** completed the **${questData.questName}** quest!`)
      .setColor(0x00ff00)
      .addFields(
        { name: '💰 Reward', value: `${questData.reward} coins`, inline: true },
        { name: '💳 New Balance', value: `${questData.newBalance} coins`, inline: true }
      )
      .setTimestamp();

    await (channel as any).send({ embeds: [embed] });
  } catch (error) {
    console.error('Error sending quest notification:', error);
  }
}

// Boost notification function
export async function sendBoostNotification(guildId: string, userTag: string, started: boolean) {
  try {
    const settings = guildSettings.get(guildId);
    if (!settings?.boostChannelId) return;

    const channel = await client.channels.fetch(settings.boostChannelId);
    if (!channel || !('send' in channel)) return;

    const embed = new EmbedBuilder()
      .setTitle(started ? '🚀 Server Boosted!' : '😢 Boost Ended')
      .setDescription(`**${userTag}** ${started ? 'started boosting' : 'stopped boosting'} the server!`)
      .setColor(started ? 0xff69b4 : 0x8b949e)
      .setTimestamp();

    await (channel as any).send({ embeds: [embed] });
  } catch (error) {
    console.error('Error sending boost notification:', error);
  }
}

// Login the bot
if (process.env.BOT2_TOKEN) {
  client.login(process.env.BOT2_TOKEN);
} else {
  console.error('❌ BOT2_TOKEN environment variable is not set');
}

export default client;

File: server/quest_bot.py

import os
from datetime import datetime
import discord
from discord import app_commands
from discord.ext import commands
import mysql.connector
from mysql.connector import pooling
from dotenv import load_dotenv

load_dotenv()

# Database connection configuration
DB_CONFIG = {
    'host': os.getenv('DB_HOST', 'mysql.railway.internal'),
    'user': os.getenv('DB_USER', 'root'),
    'password': os.getenv('DB_PASSWORD', 'qJZsyZmXvSNUxeYjsrZIKTIyAuvFMCMk'),
    'database': os.getenv('DB_NAME', 'railway'),
    'port': int(os.getenv('DB_PORT', '3306'))
}

# Create connection pool
try:
    db_pool = pooling.MySQLConnectionPool(
        pool_name="quest_bot_pool",
        pool_size=3,
        **DB_CONFIG
    )
    print("✅ Quest Bot connected to MySQL database")
except Exception as e:
    print(f"❌ Failed to connect to database: {e}")
    db_pool = None

# Bot setup
intents = discord.Intents.default()
intents.guilds = True
intents.members = True
intents.message_content = True

bot = commands.Bot(command_prefix='!', intents=intents)

# Guild settings storage
guild_settings = {}

# Database helper
def get_db_connection():
    if db_pool:
        return db_pool.get_connection()
    return None

@bot.event
async def on_ready():
    print(f'✅ Quest Bot logged in as {bot.user}!')
    
    try:
        synced = await bot.tree.sync()
        print(f'Synced {len(synced)} quest bot command(s)')
    except Exception as e:
        print(f'Failed to sync quest bot commands: {e}')

# Slash commands
@bot.tree.command(name="setquestchannel", description="Set the channel for quest completion notifications")
@app_commands.describe(channel="Channel to send quest notifications")
async def setquestchannel(interaction: discord.Interaction, channel: discord.TextChannel):
    if not interaction.user.guild_permissions.manage_channels:
        await interaction.response.send_message('❌ You need **Manage Channels** permission.', ephemeral=True)
        return
    
    guild_id = interaction.guild.id
    if guild_id not in guild_settings:
        guild_settings[guild_id] = {}
    
    guild_settings[guild_id]['questClaimChannelId'] = channel.id
    
    embed = discord.Embed(
        title='✅ Quest Channel Set',
        description=f'Quest completion notifications will now be sent to {channel.mention}',
        color=0x00ff00,
        timestamp=datetime.now()
    )
    
    await interaction.response.send_message(embed=embed)

@bot.tree.command(name="setboostchannel", description="Set the channel for boost notifications")
@app_commands.describe(channel="Channel to send boost notifications")
async def setboostchannel(interaction: discord.Interaction, channel: discord.TextChannel):
    if not interaction.user.guild_permissions.manage_channels:
        await interaction.response.send_message('❌ You need **Manage Channels** permission.', ephemeral=True)
        return
    
    guild_id = interaction.guild.id
    if guild_id not in guild_settings:
        guild_settings[guild_id] = {}
    
    guild_settings[guild_id]['boostChannelId'] = channel.id
    
    embed = discord.Embed(
        title='✅ Boost Channel Set',
        description=f'Server boost notifications will now be sent to {channel.mention}',
        color=0x00ff00,
        timestamp=datetime.now()
    )
    
    await interaction.response.send_message(embed=embed)

@bot.tree.command(name="removequestchannel", description="Remove quest notification channel")
async def removequestchannel(interaction: discord.Interaction):
    if not interaction.user.guild_permissions.manage_channels:
        await interaction.response.send_message('❌ You need **Manage Channels** permission.', ephemeral=True)
        return
    
    guild_id = interaction.guild.id
    settings = guild_settings.get(guild_id, {})
    
    if 'questClaimChannelId' not in settings:
        await interaction.response.send_message('❌ No quest channel is currently set.', ephemeral=True)
        return
    
    del settings['questClaimChannelId']
    guild_settings[guild_id] = settings
    
    embed = discord.Embed(
        title='✅ Quest Channel Removed',
        description='Quest completion notifications have been disabled.',
        color=0xff6b6b,
        timestamp=datetime.now()
    )
    
    await interaction.response.send_message(embed=embed)

@bot.tree.command(name="removeboostchannel", description="Remove boost notification channel")
async def removeboostchannel(interaction: discord.Interaction):
    if not interaction.user.guild_permissions.manage_channels:
        await interaction.response.send_message('❌ You need **Manage Channels** permission.', ephemeral=True)
        return
    
    guild_id = interaction.guild.id
    settings = guild_settings.get(guild_id, {})
    
    if 'boostChannelId' not in settings:
        await interaction.response.send_message('❌ No boost channel is currently set.', ephemeral=True)
        return
    
    del settings['boostChannelId']
    guild_settings[guild_id] = settings
    
    embed = discord.Embed(
        title='✅ Boost Channel Removed',
        description='Server boost notifications have been disabled.',
        color=0xff6b6b,
        timestamp=datetime.now()
    )
    
    await interaction.response.send_message(embed=embed)

@bot.tree.command(name="questsettings", description="View current quest notification settings")
async def questsettings(interaction: discord.Interaction):
    guild_id = interaction.guild.id
    settings = guild_settings.get(guild_id, {})
    
    quest_channel_id = settings.get('questClaimChannelId')
    boost_channel_id = settings.get('boostChannelId')
    prefix = settings.get('prefix', '!')
    
    embed = discord.Embed(
        title='🎯 Quest Bot Settings',
        color=0x7c3aed,
        timestamp=datetime.now()
    )
    
    embed.add_field(
        name='🎉 Quest Channel',
        value=f'<#{quest_channel_id}>' if quest_channel_id else 'Not set',
        inline=True
    )
    embed.add_field(
        name='🚀 Boost Channel',
        value=f'<#{boost_channel_id}>' if boost_channel_id else 'Not set',
        inline=True
    )
    embed.add_field(
        name='⚙️ Command Prefix',
        value=prefix,
        inline=True
    )
    
    await interaction.response.send_message(embed=embed)

@bot.tree.command(name="setprefix", description="Set command prefix for page information commands")
@app_commands.describe(prefix="New command prefix (e.g., '!')")
async def setprefix(interaction: discord.Interaction, prefix: str):
    if not interaction.user.guild_permissions.manage_guild:
        await interaction.response.send_message('❌ You need **Manage Server** permission.', ephemeral=True)
        return
    
    if len(prefix) > 3:
        await interaction.response.send_message('❌ Prefix must be 3 characters or less.', ephemeral=True)
        return
    
    guild_id = interaction.guild.id
    if guild_id not in guild_settings:
        guild_settings[guild_id] = {}
    
    guild_settings[guild_id]['prefix'] = prefix
    
    embed = discord.Embed(
        title='✅ Prefix Updated',
        description=f'Command prefix is now set to: `{prefix}`\n\nTry: `{prefix}page home` or `{prefix}page servers`',
        color=0x00ff00,
        timestamp=datetime.now()
    )
    
    await interaction.response.send_message(embed=embed)

# Message commands for page info
@bot.event
async def on_message(message):
    if message.author.bot:
        return
    
    if not message.guild:
        return
    
    guild_id = message.guild.id
    settings = guild_settings.get(guild_id, {})
    prefix = settings.get('prefix', '!')
    
    if not message.content.startswith(prefix):
        await bot.process_commands(message)
        return
    
    args = message.content[len(prefix):].strip().split()
    if not args:
        await bot.process_commands(message)
        return
    
    command = args[0].lower()
    
    if command == 'page':
        await handle_page_command(message, args[1:] if len(args) > 1 else [])
    
    await bot.process_commands(message)

async def handle_page_command(message, args):
    if not args:
        await message.reply('Usage: `!page <page_name>` (e.g., `!page home`, `!page servers`)')
        return
    
    page_name = args[0].lower()
    base_url = os.getenv('APP_URL', 'https://axiomer.up.railway.app')
    
    # Simple page info (in production, this would fetch from API)
    pages = {
        'home': {
            'title': 'Axiom - Discord Server Discovery',
            'description': 'Discover and promote your Discord servers',
            'url': '/'
        },
        'servers': {
            'title': 'Explore Servers',
            'description': 'Browse amazing Discord servers',
            'url': '/explore'
        },
        'bots': {
            'title': 'Discord Bots',
            'description': 'Find powerful Discord bots',
            'url': '/explore'
        },
        'quests': {
            'title': 'Daily Quests',
            'description': 'Complete quests to earn coins',
            'url': '/quest'
        },
        'store': {
            'title': 'Coin Store',
            'description': 'Spend your coins on boosts',
            'url': '/store'
        }
    }
    
    page_info = pages.get(page_name)
    if not page_info:
        await message.reply(f'❌ Page "{page_name}" not found. Available pages: home, servers, bots, store, quests')
        return
    
    embed = discord.Embed(
        title=f'📄 {page_info["title"]}',
        description=page_info['description'],
        color=0x7c3aed,
        timestamp=datetime.now(),
        url=f'{base_url}{page_info["url"]}'
    )
    
    await message.reply(embed=embed)

# Quest notification function
async def send_quest_notification(guild_id: int, user_id: str, quest_data: dict):
    try:
        settings = guild_settings.get(guild_id, {})
        channel_id = settings.get('questClaimChannelId')
        
        if not channel_id:
            return
        
        channel = bot.get_channel(channel_id)
        if not channel:
            return
        
        embed = discord.Embed(
            title='🎉 Quest Completed!',
            description=f'**{quest_data["userTag"]}** completed the **{quest_data["questName"]}** quest!',
            color=0x00ff00,
            timestamp=datetime.now()
        )
        embed.add_field(name='💰 Reward', value=f'{quest_data["reward"]} coins', inline=True)
        embed.add_field(name='💳 New Balance', value=f'{quest_data["newBalance"]} coins', inline=True)
        
        await channel.send(embed=embed)
    except Exception as e:
        print(f'Error sending quest notification: {e}')

# Boost notification function
async def send_boost_notification(guild_id: int, user_tag: str, started: bool):
    try:
        settings = guild_settings.get(guild_id, {})
        channel_id = settings.get('boostChannelId')
        
        if not channel_id:
            return
        
        channel = bot.get_channel(channel_id)
        if not channel:
            return
        
        embed = discord.Embed(
            title='🚀 Server Boosted!' if started else '😢 Boost Ended',
            description=f'**{user_tag}** {"started boosting" if started else "stopped boosting"} the server!',
            color=0xff69b4 if started else 0x8b949e,
            timestamp=datetime.now()
        )
        
        await channel.send(embed=embed)
    except Exception as e:
        print(f'Error sending boost notification: {e}')

# Start bot
def start_quest_bot():
    bot_token = os.getenv('BOT2_TOKEN')
    if not bot_token:
        print('❌ BOT2_TOKEN not found')
        return
    
    try:
        bot.run(bot_token)
    except Exception as e:
        print(f'❌ Failed to start quest bot: {e}')

if __name__ == '__main__':
    start_quest_bot()
